minimal `test` crate

- `term`
  - [ ] `collections::HashMap`
  - [ ] `env`
  - [ ] `io::BufReader`
  - [ ] `fs::metadata`
- `getopts`
  - [ ] ???
- `test`
  - [ ] ???

---

Let's make sure the `test` crate can be compiled against `steed` then we'll be able to use `#[test]` with `steed`.

It's OK to heavily modify the source as long as the main functionality remains. It's OK to drop parallelism and stub the `catch_unwind` calls (let panic!s kill the test runner)

Errors from compiling the `term` crate:

```
error[E0432]: unresolved import `std::collections::HashMap`
error[E0432]: unresolved import `std::env`
error[E0432]: unresolved import `std::io::BufReader`
error[E0425]: cannot find function `metadata` in module `fs`
```

Shimming `env::home_dir`, `env::var` and `env::var_os` for `libterm` and removing the `extern crate` definitions for `libc` and `panic_unwind`, the following errors are produced for `libtest` (duplicates removed):

    error[E0432]: unresolved import `std::panic::catch_unwind`
    error[E0432]: unresolved import `std::panic::AssertUnwindSafe`
    error[E0432]: unresolved import `std::sync::mpsc::channel`
    error[E0432]: unresolved import `std::sync::mpsc::Sender`
    error[E0432]: unresolved import `std::sync::Mutex`
    error[E0432]: unresolved import `std::thread`
    error[E0432]: unresolved import `std::sync::mpsc::RecvTimeoutError`
    error[E0425]: cannot find function `args` in module `env`
    error[E0425]: cannot find function `set_print` in module `io`
    error[E0425]: cannot find function `set_panic` in module `io`
    error: no method named `is_nan` found for type `f64` in the current scope
    error: no method named `abs` found for type `f64` in the current scope
    error: no method named `min` found for type `f64` in the current scope
    error: no method named `max` found for type `f64` in the current scope
    error: no method named `sqrt` found for type `f64` in the current scope
    error: no method named `floor` found for type `f64` in the current scope

For the `thread` module: `thread::Builder(..).name(..).spawn(..)`.

--- a/docker/qemu.sh
+++ b/docker/qemu.sh
@@ -29,7 +29,7 @@ main() {
 
     pushd $td
 
-    curl http://wiki.qemu-project.org/download/qemu-$version.tar.bz2 | \
+    curl http://download.qemu-project.org/qemu-$version.tar.bz2 | \
         tar --strip-components=1 -xj
     ./configure \
         --disable-kvm \

--- a/test/Cargo.toml
+++ b/test/Cargo.toml
@@ -1,0 +1,4 @@
+[package]
+authors = ["Jorge Aparicio <japaricious@gmail.com>"]
+name = "test"
+version = "0.1.0"

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -1,0 +1,49 @@
+pub mod test {
+    pub use ShouldPanic;
+}
+
+pub struct TestDesc {
+    pub ignore: bool,
+    pub name: StaticTestName,
+    pub should_panic: ShouldPanic,
+}
+
+pub struct TestDescAndFn {
+    pub desc: TestDesc,
+    pub testfn: StaticTestFn,
+}
+
+pub struct StaticTestFn(pub fn());
+pub struct StaticTestName(pub &'static str);
+
+pub enum ShouldPanic {
+    No,
+    Yes,
+}
+
+pub fn test_main_static(tests: &'static [TestDescAndFn]) {
+    let n = tests.len();
+
+    println!("Running {} test{}", n, if n == 1 { "" } else { "s" });
+
+    let mut passed = 0;
+    let mut ignored = 0;
+    for test in tests {
+        match test.desc.should_panic {
+            ShouldPanic::No => {
+                print!("test {} ... ", test.desc.name.0);
+                test.testfn.0();
+                println!("ok");
+                passed += 1;
+            }
+            ShouldPanic::Yes => {
+                println!("test {} ... ignored", test.desc.name.0);
+                ignored += 1;
+            }
+        }
+    }
+
+    println!("\ntest result: ok. {} passed; 0 failed; {} ignored; 0 measured",
+             passed,
+             ignored);
+}

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -24,6 +24,10 @@ pub struct StaticTestName(pub &'static str);
 
 pub struct Bencher {}
 
+impl Bencher {
+    pub fn iter<T, F>(&mut self, _: F) where F: FnMut() -> T {}
+}
+
 #[derive(Clone, Copy)]
 pub enum ShouldPanic {
     No,

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -1,3 +1,5 @@
+pub use self::TestFn::*;
+
 pub mod test {
     pub use ShouldPanic;
 }
@@ -10,12 +12,19 @@ pub struct TestDesc {
 
 pub struct TestDescAndFn {
     pub desc: TestDesc,
-    pub testfn: StaticTestFn,
+    pub testfn: TestFn,
 }
 
-pub struct StaticTestFn(pub fn());
+pub enum TestFn {
+    StaticBenchFn(fn(&mut Bencher)),
+    StaticTestFn(fn()),
+}
+
 pub struct StaticTestName(pub &'static str);
 
+pub struct Bencher {}
+
+#[derive(Clone, Copy)]
 pub enum ShouldPanic {
     No,
     Yes,
@@ -29,14 +38,14 @@ pub fn test_main_static(tests: &'static [TestDescAndFn]) {
     let mut passed = 0;
     let mut ignored = 0;
     for test in tests {
-        match test.desc.should_panic {
-            ShouldPanic::No => {
+        match (test.desc.should_panic, &test.testfn) {
+            (ShouldPanic::No, &TestFn::StaticTestFn(ref f)) => {
                 print!("test {} ... ", test.desc.name.0);
-                test.testfn.0();
+                f();
                 println!("ok");
                 passed += 1;
             }
-            ShouldPanic::Yes => {
+            _ => {
                 println!("test {} ... ignored", test.desc.name.0);
                 ignored += 1;
             }

--- a/test/src/lib.rs
+++ b/test/src/lib.rs
@@ -30,6 +30,8 @@ pub enum ShouldPanic {
     Yes,
 }
 
+pub fn black_box<T>(_: T) {}
+
 pub fn test_main_static(tests: &'static [TestDescAndFn]) {
     let n = tests.len();
 

